datafusion/common/src/dfschema.rs

```diff
@@ -28,18 +29,58 @@ use crate::{field_not_found, Column, OwnedTableReference, TableReference};
 
 use arrow::compute::can_cast_types;
 use arrow::datatypes::{DataType, Field, FieldRef, Fields, Schema, SchemaRef};
-use std::fmt::{Display, Formatter};
 
 /// A reference-counted reference to a `DFSchema`.
 pub type DFSchemaRef = Arc<DFSchema>;
 
+/// Stores identifier keys and their associated column indices. An identifier key
+/// is a column whose value determines values of some other (dependent) columns.
+/// These dependent columns are the "associated columns" of this identifier key.
+/// If two rows have the same identifier key, associated columns in these rows
+/// are necessarily the same. If the identifier key is unique, the set of
+/// associated columns is equal to the entire schema and the identifier key can
+/// serve as a primary key. Note that a primary key may "downgrade" into an
+/// identifier key due to an operation such as a join, and this object is used to
+/// track dependence relationships in such cases.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct IdentifierKeyGroup {
+    pub identifier_key_indices: Vec<usize>,
+    /// Flag indicating whether this identifier key is unique.
+    /// If true, these indices constitute a primary key. Otherwise, the
+    /// identifier key still uniquely determines its associated columns.
+    pub is_unique: bool,
+    pub associated_indices: Vec<usize>,
+}
+
+impl IdentifierKeyGroup {
+    pub fn new(
+        identifier_key_indices: Vec<usize>,
+        associated_indices: Vec<usize>,
+    ) -> Self {
+        Self {
+            identifier_key_indices,
+            is_unique: false,
+            associated_indices,
+        }
+    }
+
+    pub fn with_is_unique(mut self, is_unique: bool) -> Self {
+        self.is_unique = is_unique;
+        self
+    }
+}
+
+pub type IdentifierKeyGroups = Vec<IdentifierKeyGroup>;
```

Comment:
Stylistically if you made this a real type, it would:
1. encapsulate the manipulation more (like making `add_offset_to_identifier_key_groups` a method)
2. Make finding the available operations on it easier in the future

Something like


```suggestion
pub struct IdentifierKeyGroups {
  inner:  Vec<IdentifierKeyGroup>;
}

impl IdentifierKeyGroups {
  pub add_offset(&mut self, offset: usize) {..}
  ...
}
```